Shreevaths taking over for Manvitha teacher request resources

myeeli opened the pull request:
# Description
<img width="583" height="274" alt="image" src="https://github.com/user-attachments/assets/c2e1b79b-4529-4e79-9a12-ebfa8a81f54e" />

<img width="556" height="233" alt="image" src="https://github.com/user-attachments/assets/63ebbba6-ff66-4be8-a18c-1a8d46b08a61" />


## Related PRS (if any):
Previous PR: https://github.com/OneCommunityGlobal/HGNRest/pull/1907

## Main changes explained:
- Implemented a secure educator–PM resource request system with proper role validation.
- Removed requestor_id and cleaned database schema for accuracy and simplicity.
- Fixed authorization logic so only educators create requests and only PMs update statuses.
- Added strict ObjectId validation and prevented educators from overriding request status.
- Improved controller structure, error handling, and added pagination for PM dashboards.
- Updated unit tests to cover authorization, status protection, and request validation.
…

## How to test:
- Check into the current branch.
- Log in as Educator
- all the endpoint using Postman:
   POST /educator/resource-requests
   Body:
  {
   "request_title": "Classroom Supplies",
   "request_details": "Need whiteboards and markers"
   }
   Verify: Resource request is created successfully and Status is set to pending
- Call the endpoint using Postman:
    GET /educator/resource-requests
    Verify: Educator can view their own resource requests
- Log in as a PM
- Call the endpoint using Postman: GET /pm/resource-requests
    Verify: PM can view all educator resource requests
- Call the endpoint using Postman: PUT /pm/resource-requests/:id
   Body : 
   {
   "status": "approved"
   }
  Verify: Request status is updated successfully


## Screenshots or videos of changes:

## Note:
Include the information the reviewers need to know.

saishekhar10 committed:
feat: add resource request models and controller

saishekhar10 committed:
feat: updated resoruceRequest schema

myeeli committed:
Educator role check

myeeli committed:
Educator role endpoint check

myeeli committed:
Implemented test file and router file

myeeli committed:
Permission Handling Fix

sayali-2308 reviewed (commented):
Cannot test endpoints - depends on incomplete PR #1907. All endpoints return 401 Unauthorized. PR #1940 description states 'the remaining part is to check if the user accessing the above endpoints is a PM or an Educator (PR #1907)'. Checked PR #1907 which has labels 'Do Not Review' and 'Needs New Developer'. PR #1907 note states: 'Endpoints have been created, the remaining part is to add a check if the user is an educator or a pm'. Authorization middleware not implemented in either PR. Cannot proceed with functional testing until PR #1907 is completed with proper role-based access control (educator vs PM validation). PR #1940 is blocked waiting for PR #1907 dependency.

SharadhaKasiviswanathan committed:
Resolve merge conflicts with development branch

SharadhaKasiviswanathan committed:
Fix resourceRequestController tests: proper constructor mock, populate chain, valid ObjectId

SharadhaKasiviswanathan committed:
fix(security): resolve SonarCloud E Security Rating on new code

- Fix auth check: use requestor.requestorId (matches JWT middleware shape)
  instead of requestor._id which was always undefined in production
- Fix requestor access: use req.body.requestor (codebase convention)
  instead of non-standard req.requestor || req.body.requestor pattern
- Add express-validator middleware for input validation on all routes
- Validate status query parameter against allowed enum values
- Add request_title/request_details length limits and sanitization
- Cap pagination limit at 100 to prevent resource exhaustion
- Remove console.error with raw error objects (info leak risk)
- Fix roleChecks.js: correct PM role name to 'Program Manager'
- Remove tracked .DS_Store files
- Expand test coverage: add tests for 401/403/400 edge cases

SharadhaKasiviswanathan committed:
fix(security): resolve SonarCloud BLOCKER vulnerabilities (jssecurity:S5147)

Sanitize all user-controlled values before they reach MongoDB queries:

- Add sanitizeStatus() helper: validates against ALLOWED_STATUSES and
  returns String() to break taint flow from req.query.status
- Add sanitizeObjectId() helper: validates with isValid() then creates
  new mongoose.Types.ObjectId() to break taint flow from
  req.query.educator_id and requestor.requestorId
- Apply sanitizers in all controller methods
- Add 3 new tests: status filtering, educator_id filtering, invalid status

SharadhaKasiviswanathan committed:
fix(security): use codebase mongoQuerySanitizer for SonarCloud S5147

Replace custom sanitizers with the existing mongoQuerySanitizer utility
(sanitizeObjectIdQuery, sanitizeQueryString) that already passes
SonarCloud analysis in other controllers (collaborationController,
questionSetController).

- Use sanitizeObjectIdQuery() for all ObjectId params (requestorId,
  req.params.id, req.query.educator_id) - returns validated string
- Use sanitizeQueryString() + ALLOWED_STATUSES whitelist for status
- Remove custom sanitizeObjectId/sanitizeString helpers
- Remove mongoose import (no longer needed in controller)
- Update test expectations for string-based filter values

SharadhaKasiviswanathan committed:
fix(code-review): resolve SonarCloud MINOR issues

- Use Set for ALLOWED_STATUSES and .has() instead of Array.includes()
- Add console.log to catch blocks to handle exceptions properly
- Name anonymous router export function resourceRequestRoutes

SharadhaKasiviswanathan committed:
ci: add CI-based SonarCloud analysis to resolve automatic analysis timeout

Automatic Analysis timed out on this large Node.js codebase.
Switch to CI-based analysis using SonarSource/sonarqube-scan-action
which runs in GitHub Actions with self-managed hardware.

- Add sonar-project.properties (projectKey, org, sources, exclusions)
- Add SonarCloud job to node.js.yml workflow (runs after lint)
- Uses SONAR_TOKEN secret for authentication

Requires SONAR_TOKEN to be configured in repo Settings > Secrets

SharadhaKasiviswanathan committed:
fix(ci): pin SonarCloud action to full SHA and make conditional on SONAR_TOKEN

- Pin SonarSource/sonarqube-scan-action to full commit SHA
  (689fb39b...) to resolve SonarCloud VULNERABILITY S6SR
- Make sonarcloud job conditional: skipped when SONAR_TOKEN
  is not configured, so it doesn't block PRs

SharadhaKasiviswanathan committed:
fix(ci): remove CI-based SonarCloud job (no SONAR_TOKEN configured)

The CI-based SonarCloud job requires a SONAR_TOKEN secret that isn't
configured. Remove it to restore passing CI. SonarCloud analysis is
handled by the GitHub App automatic analysis instead.

The sonar-project.properties file is retained for the GitHub App.

SharadhaKasiviswanathan committed:
fix: restore accidentally removed routes and remove sonar-project.properties

- Restore materialCostRouter and lessonPlanSubmissionRouter routes
  in routes.js (accidentally deleted during merge)
- Remove sonar-project.properties (was not present on development;
  SonarCloud automatic analysis handles project detection natively)

kzou55 reviewed (dismissed):
Hi Sharadha, 

I ran and tested the PR locally

## Verified
- Making a request
    <img width="1424" height="635" alt="Image" src="https://github.com/user-attachments/assets/6b655924-155e-42a7-bce2-58931bde6e35" />

- Getting the requests

    <img width="1424" height="524" alt="Image" src="https://github.com/user-attachments/assets/770f8b95-5d50-487c-ad83-3ca841a360eb" />

## Issues/Clarification
1. By PM, do you just mean any account with high level of access like an admin? I don't see a PM role when creating a test user. So I just used my Admin Account. Could you clarify? As the following requests work when using an Admin account
    - Getting educator requests
    <img width="1429" height="974" alt="Image" src="https://github.com/user-attachments/assets/032e47d9-7373-430e-a38d-36adc01e742a" />

    - PUT request approved works but I did it as an Admin

    
    <img width="1417" height="681" alt="Image" src="https://github.com/user-attachments/assets/3a35e2cf-855f-4149-a48c-93e4486e2c01" />

    <img width="1433" height="961" alt="Image" src="https://github.com/user-attachments/assets/b572e7da-c94a-46ac-885f-bf9dc707b449" />

akv-iu reviewed (approved):
## Verified
- Every Functionality works as expected and API end points, work correctly.
- Tested all the functionality, added few samples. 

# Samples

<img width="2306" height="1534" alt="Image" src="https://github.com/user-attachments/assets/74ba7805-ba44-459b-8523-0e7b3ed45e71" />

- Works when the account doesn't have a desired role

<img width="2292" height="1400" alt="Image" src="https://github.com/user-attachments/assets/08ad0c18-5ad1-4fa9-8920-5000ca792a1f" />

shreevathsrao committed:
fix: resolve merge conflict

shreevathsrao committed:
fix(merge): Merge branch development into Manvitha-Teacher-Request-Resources

shreevathsrao committed:
fix(remove): remove unwanted files

shreevathsrao committed:
fix(remove): rmv unwanted file changes

shreevathsrao committed:
fix(add): Write test cases

iAbhi001 reviewed (approved):
Thanks for sharing these details!
The permission checks look clear:
Educator requests: Requires Educator role OR createResourceRequests permission.
PM requests: Requires Program Manager role, manageResourceRequests permission, or Owner/Administrator roles (with status updates additionally accepting updatePMResourceRequestStatus).
I’ll dive into this and let you know if I need any further clarification.


<img width="1280" height="832" alt="Image" src="https://github.com/user-attachments/assets/a2683918-352c-43a3-a3fc-98f41b530dc8" />

<img width="1280" height="832" alt="Image" src="https://github.com/user-attachments/assets/c6026e55-7e3b-4fce-b86a-99adb69cfe60" />

<img width="1280" height="832" alt="Image" src="https://github.com/user-attachments/assets/19afe13e-eeb5-4295-9733-610fd10020e9" />

<img width="1280" height="832" alt="Image" src="https://github.com/user-attachments/assets/9e1ff68a-17db-4fcc-8362-07e96af60e94" />

DeepighaJ reviewed (approved):
Tested the PR in Postman by creating request as Educator and verified/updated as Owner account. The functionalities working as expected.

**Logged in as Educator:**

- Verified Resource request is created successfully and Status is set to pending

<img width="910" height="526" alt="Image" src="https://github.com/user-attachments/assets/b707dbd8-26d2-47ca-b643-9eab565d567c" />

- Verified Educator can view their own request

<img width="912" height="437" alt="Image" src="https://github.com/user-attachments/assets/e017d5e7-6c87-4c4d-acc2-f571de7c7f18" />

**Logged in as PM:**

- Verified PM can view all educator resource requests

<img width="920" height="620" alt="Image" src="https://github.com/user-attachments/assets/61257015-559c-478e-a4c0-d25618d02330" />

- Verified Request status is updated to Approved 

<img width="922" height="627" alt="Image" src="https://github.com/user-attachments/assets/a08f0d4b-928b-48e1-a4d8-fecbc7061f35" />
